File: python/OneKnownFamily.py
from __future__ import annotations

# coding: utf-8
# OneKnownFamily.py
# 片親がphasingされていないがknownでもう片親はunknownな家系を補完する

from typing import Optional

from VCF import VCFSmall
from VCFGeno import VCFGenoBase, VCFGeno
from VCFFamily import *
from VCFImputable import VCFImputable
from Map import *
from KnownFamily import *
from VCFOneParentKnown import VCFOneParentKnown
from ReferenceHaplotype import filter_haplotypes
from OptionSmall import OptionSmall


# HMMにrefを使っても計算量が十分に小さいか
def is_small(ref_haps: list[list[int]], L: int, op: OptionSmall) -> bool:
	M = len(ref_haps[0])				# マーカー数
	NH = len(ref_haps)
	R = NH**2 * (2*NH - 1) / op.precision_ratio
	return R * M < 10**8 and R < 10**5 and L * R * M < 10**9

# is_smallが通るギリギリのNHを求める
def compute_upper_NH(family: Family, M: int, L: int, op: OptionSmall) -> int:
	for NH in count(2):
		R = NH**2 * (2*NH - 1) / op.precision_ratio
		if not (R * M < 10**8 and R < 10**5 and L * R * M < 10**9):
			break
	return NH - 1

def create_family_vcf(family: KnownFamily, records: list[VCFFamilyRecord],
					num_families: int, ref_haps: list[list[int]],
					vcf: VCFSmall, op: OptionSmall) -> Optional[VCFImputable]:
	NH = compute_upper_NH(family, len(vcf), num_families, op)
	if is_small(ref_haps, num_families, op):
		return VCFOneParentKnown(family.samples(), records, ref_haps,
											family.mat_known, op.map, vcf)
	elif NH >= 10:
		NH2 = min(20, NH)
		if family.mat_known:
			gts = [ r.geno[0] for r in records ]
			ref_haps = filter_haplotypes(ref_haps, gts, NH2)
		else:
			gts = [ r.geno[1] for r in records ]
			ref_haps = filter_haplotypes(ref_haps, gts, NH2)
		return VCFOneParentKnown(family.samples(), records, ref_haps,
											family.mat_known, op.map, vcf)
	else:
		return None

def impute(orig_vcf: VCFSmall, imputed_vcf: VCFGeno, ref_haps: list[list[int]],
									families: list[KnownFamily],
									op: OptionSmall) -> Optional[VCFGenoBase]:
	vcfs: list[VCFGenoBase] = []
	for family in families:
		vcf = VCFFamily.create_by_two_vcfs(imputed_vcf,
											orig_vcf, family.samples())
		parent = vcf.mat() if family.mat_known else vcf.pat()
		vcf1 = create_family_vcf(family, vcf.records, len(families),
													ref_haps, orig_vcf, op)
		if vcf1 is None:
			continue
		vcf1.impute()
		vcf1_parent = vcf1.extract_by_samples([parent])
		vcfs.append(vcf1_parent)
	
	if not vcfs:
		return None
	
	print("%d families whose one parent is known and the other parent is"
								" unknown have been imputed." % len(families))
	new_vcf = VCFGeno.join(vcfs, orig_vcf.samples)
	return new_vcf

__all__ = ['impute']

File: src/OneKnownFamily.cpp
#include <algorithm>
#include <cassert>
#include "../include/OneKnownFamily.h"
#include "../include/VCFOneParentKnown.h"
#include "../include/KnownFamily.h"
#include "../include/ReferenceHaplotype.h"
#include "../include/OptionSmall.h"
#include "../include/common.h"

using namespace std;


//////////////////// OneKnownFamily ////////////////////

// Is the computational cost sufficiently small even when using ref in HMM?
bool OneKnownFamily::is_small(const vector<vector<int>>& ref_haps,
											size_t L, const OptionSmall& op) {
	const size_t	M = ref_haps[0].size();
	const size_t	NH = ref_haps.size();
	if(NH >= 100)
		return false;
	
	const double	R = (NH * NH * (2*NH - 1)) / op.precision_ratio;
	return R*M < 1e8 && R < 1e5 && L*R*M < 1e9;
}

// upper NH which passes is_small
size_t OneKnownFamily::compute_upper_NH(const Family *family, size_t M,
										size_t L, const OptionSmall& op) {
	for(size_t NH = 2; ; ++NH) {
		const double	R = (NH * NH * (2*NH - 1)) / op.precision_ratio;
		if(!(R * M < 1e8 && R < 1e5 && L * R * M < 1e9))
			return NH - 1;
	}
	return 0;	// dummy
}

VCFImputable *OneKnownFamily::create_family_vcf(
									const KnownFamily *family,
									const vector<VCFFamilyRecord *>& records,
									size_t num_families,
									const vector<vector<int>>& ref_haps,
									const VCFSmall *orig_vcf,
									const OptionSmall& op) {
	const size_t	lower_NH = 10;
	const size_t	upper_NH = 20;
	const bool	is_mat_known = family->is_mat_known();
	const size_t	M = ref_haps[0].size();
	const size_t	NH = compute_upper_NH(family, M, num_families, op);
	if(is_small(ref_haps, num_families, op)) {
		return new VCFOneParentKnown(family->get_samples(),
										records, ref_haps, is_mat_known,
										op.map, 0.01, orig_vcf);
	}
	else if(NH >= lower_NH) {
		const size_t	NH2 = min(upper_NH, NH);
		const size_t	known_index = is_mat_known ? 0 : 1;
		const auto	gts = VCFFamilyRecord::extract_sample_genotypes(
														known_index, records);
		const auto	ref_haps1 =
				ReferenceHaplotype::filter_haplotypes(ref_haps, gts, NH2);
		return new VCFOneParentKnown(family->get_samples(), records,
											ref_haps1, is_mat_known,
											op.map, 0.01, orig_vcf);
	}
	else {
		return NULL;
	}
}

VCFGenoBase *OneKnownFamily::impute(const VCFSmall *orig_vcf,
									const VCFGeno *imputed_vcf,
									const vector<vector<int>>& ref_haps,
									const vector<const KnownFamily *>& families,
									const OptionSmall& op) {
	const size_t	N = families.size();
	vector<VCFImputable *>	vcfs;
	vector<VCFFamily *>	vcf_garbage;
	for(size_t i = 0; i < N; ++i) {
		const KnownFamily	*family = families[i];
		auto	*vcf = VCFFamily::create_by_two_vcfs(imputed_vcf,
											orig_vcf, family->get_samples());
		auto	*vcf1 = create_family_vcf(family, vcf->get_family_records(),
													N, ref_haps, orig_vcf, op);
		if(vcf1 != NULL) {
			vcfs.push_back(vcf1);
			vcf->clear_records();
		}
		vcf_garbage.push_back(vcf);
	}
	
	if(vcfs.empty()) {
		Common::delete_all(vcf_garbage);
		return NULL;
	}
	
	VCFImputable::impute_VCFs(vcfs, op.num_threads);
	
	cout << vcfs.size() << " families whose one parent is known and"
				<< " the other parent is unknown have been imputed." << endl;
	
	// Small VCFs are heavy to process, so it will be parallelized.
	vector<const VCFGenoBase *>	vcfs1(vcfs.begin(), vcfs.end());
	auto	*new_vcf = VCFGeno::join(vcfs1, orig_vcf->get_samples());
	Common::delete_all(vcf_garbage);
	Common::delete_all(vcfs);
	return new_vcf;
}
